modules/javafx.graphics/src/main/java/javafx/css/Selector.java

```diff
@@ -150,9 +173,10 @@ protected void writeBinary(DataOutputStream os, StyleConverter.StringStore strin
      * @param bssVersion bss version identifier
      * @param is {@code DataInputStream} to read {@code Selector} data from
      * @param strings string array containing selector details
+     * @return a selector, never {@code null}
      * @throws IOException if reading from {@code DataInputStream} fails
      */
-    static Selector readBinary(int bssVersion, DataInputStream is, String[] strings)
+    protected static Selector readBinary(int bssVersion, DataInputStream is, String[] strings)
```

Comment:
This is an API addition, so will need an `@since 23`. Since you are adding new API here, should it be an instance method rather than a static method? Or is is called without having an instance of Selector in some cases?

Reply:
I'll add the `@since 23`, however it can't be called by anyone except FX itself.

Some background: the readBinary/writeBinary methods are ultimately called via the publicly accessible methods `loadBinary` and `saveBinary` in `javafx.css.Stylesheet`.  

The reason it needs to be `protected` now is because `CompoundSelector` is using this (but IMHO, it shouldn't have).  Why I say it shouldn't?  That's because it already knows what it will be reading will all be `SimpleSelector`s, as it stores a counter (a `short`) and then loads that many `SimpleSelector`s.  However, by not taking the direct route of using `SimpleSelector#readBinary` (and the corresponding `SimpleSelector#writeBinary`) there is an additional `byte` prefix indicating the type of selector -- this isn't needed and wastes some space in the binary format.

Changing that now however would make the binary format incompatible with earlier versions, so I think making this `protected` is a reasonable solution.  It also mirrors the `writeBinary` method then, which was already `protected`.

Reply:
I missed something in my above evaluation.  The counterpart method `writeBinary` is not `static`.  Although that makes sense as in that case you do have an instance you wish to write, it does make the read/write API asymmetric.

It's not possible to make `readBinary` an instance method though as it is the method that is creating those instances.

The other way around is possible (make `writeBinary` static), but it would then again need a pattern match to determine the correct static `writeBinary` to call when writing an arbitrary `Selector`.  However, this would have allowed `CompoundSelector` to call a static version of `SimpleSelector#writeBinary` and `readBinary`, avoiding the extra identifying byte in the binary format, and avoiding the cast when reading it back.

The read/write loops below:

+
        final int nSelectors = is.readShort();
        final List<SimpleSelector> selectors = new ArrayList<>();
        for (int n=0; n<nSelectors; n++) {
            selectors.add((SimpleSelector)Selector.readBinary(bssVersion, is,strings));
        }
+
        os.writeShort(selectors.size());   // writes the number of SimpleSelectors to the binary stream
        for (int n=0; n< selectors.size(); n++) selectors.get(n).writeBinary(os,stringStore);  // writes each individually

Would then have become:

+
        final int nSelectors = is.readShort();
        final List<SimpleSelector> selectors = new ArrayList<>();
        for (int n=0; n<nSelectors; n++) {
            selectors.add(SimpleSelector.readBinary(bssVersion, is,strings));  // cast is gone
        }
+
        os.writeShort(selectors.size());   // writes the number of SimpleSelectors to the binary stream
        for (int n=0; n< selectors.size(); n++) SimpleSelector.writeBinaryStatic(selectors.get(n), os, stringStore);  // writes each individually

Reply:
it looks to me readBinary should be static, and writeBinary an instance method - this is a normal pattern.  the asymmetry is dictated by the fact that we don't have an instance when reading, so typically read() methods read the stream and create an instance at the last moment with the specific constructor.  unless, of course, the design allows for mutation and the fields can be set().

Alternatively, readBinary() could be moved to another class (helper? reader?) to avoid making this an API change.

what do you think?

Reply:
I can see if I can externalize this or if that would run into issues.  Also please note, although technically an API change, it is NOT an accessible API (and so can be removed at any time) because only the permitted types can access this API.

In other words, this could wait and be done separately or never.

Reply:
moving it to an internal class would eliminate the public API.

I agree with you that the risk is rather low because because the probability of people ever touching these classes is nearly zero.

Reply:
I am fine with either approach. If it is easy, you can move it now, but there are no compatibility concerns with doing it later since a protected method of a sealed class without any exported subclasses is not accessible.

Reply:
what is the resolution, @hjohn ?

Reply:
- I've moved the serialization logic to `BinarySerializer` and removed the two methods from `Selector`.  `writeBinary` was `protected` but couldn't be accessed before anyway as it was impossible to subclass `Selector` (as it had a package private constructor)
- I've moved most of the code to create a `Match` to the `Match` class so the relevant fields can now be `private`. 

Let me know if this is an improvement.